brushfire-core/src/main/scala/com/stripe/brushfire/Brushfire.scala

```diff
@@ -46,10 +46,9 @@ trait Split[V, T] {
   def predicates: Iterable[(Predicate[V], T)]
 }
 
-/** Evaluates the goodness of a candidate split */
-trait Evaluator[V, T] {
-  /** returns a (possibly transformed) version of the input split, and a numeric goodness score */
-  def evaluate(split: Split[V, T]): (Split[V, T], Double)
+trait Evaluator[T] {
+  /** returns an overall numeric training error for a tree or split, or None for infinite/unacceptable error */
+  def trainingError(root: T, leaves: Iterable[T]): Option[Double]
```

Comment:
So, I think we should stick with `Double` instead of `Option[Double]`. My guess is that a `NaN` value would be totally fine to use for this purpose (it certainly won't be a valid score).

The idea of using some `Double => Boolean` values to test criteria seems fine to me. I just think we don't need to create more boxes in this case.

(Sorry I am late to this party.)